src/NServiceBus.Persistence.DynamoDB/Saga/SagaPersister.cs

```diff
@@ -24,19 +27,138 @@ public SagaPersister(SagaPersistenceConfiguration configuration, IAmazonDynamoDB
 
         public async Task<TSagaData> Get<TSagaData>(Guid sagaId, ISynchronizedStorageSession session, ContextBag context, CancellationToken cancellationToken = default) where TSagaData : class, IContainSagaData
         {
-            var getItemRequest = new GetItemRequest
+            if (configuration.UsePessimisticLocking)
+            {
+                return await ReadWithLock<TSagaData>(sagaId, context, session, cancellationToken).ConfigureAwait(false);
+            }
+            else
             {
-                ConsistentRead = true,
-                Key = new Dictionary<string, AttributeValue>
+                // Using optimistic concurrency control
+                var getItemRequest = new GetItemRequest
+                {
+                    ConsistentRead = true,
+                    Key = new Dictionary<string, AttributeValue>
                     {
                         { configuration.PartitionKeyName, new AttributeValue { S = $"SAGA#{sagaId}" } },
                         { configuration.SortKeyName, new AttributeValue { S = $"SAGA#{sagaId}" } }, //Sort key
                     },
-                TableName = configuration.TableName
-            };
+                    TableName = configuration.TableName
+                };
 
-            var response = await dynamoDbClient.GetItemAsync(getItemRequest, cancellationToken).ConfigureAwait(false);
-            return !response.IsItemSet ? default : Deserialize<TSagaData>(response.Item, context);
+                var response = await dynamoDbClient.GetItemAsync(getItemRequest, cancellationToken).ConfigureAwait(false);
+                return !response.IsItemSet ? default : Deserialize<TSagaData>(response.Item, context);
+            }
+        }
+
+        async Task<TSagaData> ReadWithLock<TSagaData>(Guid sagaId, ContextBag context,
+            ISynchronizedStorageSession synchronizedStorageSession, CancellationToken cancellationToken)
+            where TSagaData : class, IContainSagaData
+        {
+            using var timedTokenSource = new CancellationTokenSource(configuration.LeaseAcquistionTimeout);
+            using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timedTokenSource.Token);
+            cancellationToken = cts.Token;
+            while (true)
```

Comment:
let's try to replicate the cosmos approach

Reply:
switch to TimeoutException in case operation cancelled with the lease acquisition ct